File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/BitWise/findDuplicateArrayBitWise.py
from sys import stdin, stdout
rl = stdin.readline
wr = stdout.write


class BitArray:

    arr = []

    def __init__(self, n):
        self.arr = [0 for x in range((n >> 5)+1)]

    def get(self, pos):
        self.index = (pos >> 5)
        self.bitNo = (pos & 0x1f)
        return (self.arr[self.index] & (1 << self.bitNo)) != 0

    def set(self, pos):
        self.index = (pos >> 5)
        self.bitNo = (pos & 0x1f)
        self.arr[self.index] |= (1 << self.bitNo)

    def checkDuplicates(self, arr):
        for i in range(len(arr)):
            num = arr[i]-1
            if ba.get(num):
                wr(f'{num+1} ')
            else:
                ba.set(num)


arr = [10, 10, 10, 5, 58, 8, 1, 5, 3, 2, 16, 4, 4]
ba = BitArray(32000)
ba.checkDuplicates(arr)

File: Codigos estudiantes por lenguaje/Java/Bryann Valderrama/BitWise/findDuplicateArrayBitWise.java
public class Main {

    static class BitArray {

        int[] arr;

        public BitArray(int n) {
            arr = new int[(n >> 5) + 1];
        }

        boolean get(int pos) {
            int index = (pos >> 5);
            int bitNo = (pos & 0x1f);
            return (arr[index] & (1 << bitNo)) != 0;
        }

        void set(int pos) {
            int index = (pos >> 5);
            int bitNo = (pos & 0x1f);
            arr[index] |= (1 << bitNo);
        }

        static void checkDuplicates(int[] arr) {
            BitArray ba = new BitArray(32000);
            for (int i = 0; i < arr.length; i++) {
                int num = arr[i] - 1;
                if (ba.get(num)) {
                    System.out.print((num + 1) + " ");
                } else {
                    ba.set(num);
                }
            }
        }
    }

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7, 8, 9};
        BitArray.checkDuplicates(arr);
    }
}
